File: ios/u/Sources/u/Media/UMediaHandler.swift
import AVFoundation

#if canImport(Flutter)
    import Flutter
#elseif canImport(FlutterMacOS)
    import FlutterMacOS
#endif

public final class UMediaHandler: NSObject {
    private let channel: FlutterMethodChannel
    private let sessionChannel: FlutterMethodChannel
    private let messenger: FlutterBinaryMessenger
    private let registry: FlutterTextureRegistry

    private var players: [Int: UMediaPlayer] = [:]
    private var nextId = 1

    public init(messenger: FlutterBinaryMessenger, registry: FlutterTextureRegistry) {
        self.messenger = messenger
        self.registry = registry
        channel = FlutterMethodChannel(name: "u/media", binaryMessenger: messenger)
        sessionChannel = FlutterMethodChannel(name: "u/media_session", binaryMessenger: messenger)
        super.init()

        channel.setMethodCallHandler { [weak self] call, result in
            self?.handle(call: call, result: result)
        }
        sessionChannel.setMethodCallHandler { call, result in
            switch call.method {
            case "requestFocus":
                result(true)
            case "abandonFocus":
                result(nil)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }

    private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        if call.method == "isAvailable" {
            result(true)
            return
        }

        let arguments = call.arguments as? [String: Any] ?? [:]

        if call.method == "create" {
            let kind = arguments["kind"] as? String ?? "video"
            let config = arguments["config"] as? [String: Any] ?? [:]
            let id = nextId
            nextId += 1
            players[id] = UMediaPlayer(
                playerId: id,
                messenger: messenger,
                registry: registry,
                config: config,
                isVideo: kind == "video"
            )
            result(id)
            return
        }

        guard let id = arguments["id"] as? Int else {
            result(FlutterError(code: "ERROR_UNSPECIFIED", message: "Missing player id", details: nil))
            return
        }
        guard let player = players[id] else {
            result(FlutterError(code: "ERROR_NOT_FOUND", message: "Player \(id) not found", details: nil))
            return
        }

        switch call.method {
        case "open":
            player.open(
                source: arguments["source"] as? [String: Any] ?? [:],
                autoPlay: arguments["autoPlay"] as? Bool ?? false,
                resumeMs: arguments["resumeMs"] as? Int
            )
            result(nil)
        case "play":
            player.play()
            result(nil)
        case "pause":
            player.pause()
            result(nil)
        case "stop":
            player.stop()
            result(nil)
        case "seek":
            player.seek(positionMs: arguments["positionMs"] as? Int ?? 0, precise: arguments["precise"] as? Bool ?? true)
            result(nil)
        case "stepFrame":
            player.stepFrame(frames: arguments["frames"] as? Int ?? 1)
            result(nil)
        case "setSpeed":
            player.setSpeed(arguments["speed"] as? Double ?? 1.0, preservePitch: arguments["preservePitch"] as? Bool ?? true)
            result(nil)
        case "setVolume":
            player.setVolume(arguments["volume"] as? Double ?? 1.0)
            result(nil)
        case "setMuted":
            player.setMuted(arguments["muted"] as? Bool ?? false)
            result(nil)
        case "setRepeat":
            player.setRepeat(arguments["mode"] as? String)
            result(nil)
        case "selectTrack":
            player.selectTrack(trackId: arguments["trackId"] as? String ?? "", type: arguments["type"] as? String)
            result(nil)
        case "setAutoQuality":
            player.setMaxHeight(0)
            result(nil)
        case "setMaxHeight":
            player.setMaxHeight(arguments["height"] as? Int ?? 0)
            result(nil)
        case "setAudioDelay":
            result(nil)
        case "enterPip":
            result(player.enterPip(aspectRatio: arguments["aspectRatio"] as? Double ?? 16.0 / 9.0))
        case "exitPip":
            player.exitPip()
            result(nil)
        case "screenshot":
            if let data = player.screenshot() {
                result(FlutterStandardTypedData(bytes: data))
            } else {
                result(nil)
            }
        case "setNotification":
            result(nil)
        case "dispose":
            player.dispose()
            players.removeValue(forKey: id)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func dispose() {
        channel.setMethodCallHandler(nil)
        sessionChannel.setMethodCallHandler(nil)
        players.values.forEach { $0.dispose() }
        players.removeAll()
    }
}

File: android/src/main/kotlin/com/sinamn75/u/media/UMediaHandler.kt
package com.sinamn75.u.media

import android.app.Activity
import android.content.Context
import androidx.media3.common.util.UnstableApi
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.TextureRegistry

@UnstableApi
class UMediaHandler(
    private val context: Context,
    private val messenger: BinaryMessenger,
    private val textureRegistry: TextureRegistry,
) : MethodChannel.MethodCallHandler {
    private val channel = MethodChannel(messenger, "u/media")
    private val players = mutableMapOf<Int, UMediaPlayer>()
    private var nextId = 1
    private var activity: Activity? = null

    init {
        channel.setMethodCallHandler(this)
    }

    fun setActivity(value: Activity?) {
        activity = value
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
        players.values.forEach { it.dispose() }
        players.clear()
        activity = null
    }

    override fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        if (call.method == "isAvailable") {
            result.success(true)
            return
        }

        if (call.method == "create") {
            val kind = call.argument<String>("kind") ?: "video"
            val config = call.argument<Map<String, Any?>>("config") ?: emptyMap()
            val id = nextId++
            players[id] = UMediaPlayer(id, context, messenger, textureRegistry, config, kind == "video")
            result.success(id)
            return
        }

        val id = call.argument<Int>("id")
        if (id == null) {
            result.error("ERROR_UNSPECIFIED", "Missing player id", null)
            return
        }
        val player = players[id]
        if (player == null) {
            result.error("ERROR_NOT_FOUND", "Player $id not found", null)
            return
        }

        try {
            handleOnPlayer(call, player, id, result)
        } catch (error: IllegalStateException) {
            result.error("ERROR_UNSPECIFIED", error.message, null)
        } catch (error: IllegalArgumentException) {
            result.error("ERROR_UNSPECIFIED", error.message, null)
        }
    }

    private fun handleOnPlayer(
        call: MethodCall,
        player: UMediaPlayer,
        id: Int,
        result: MethodChannel.Result,
    ) {
        when (call.method) {
            "open" -> {
                player.open(
                    call.argument<Map<String, Any?>>("source") ?: emptyMap(),
                    call.argument<Boolean>("autoPlay") ?: false,
                    call.argument<Number>("resumeMs")?.toLong(),
                )
                result.success(null)
            }
            "play" -> {
                player.play()
                result.success(null)
            }
            "pause" -> {
                player.pause()
                result.success(null)
            }
            "stop" -> {
                player.stop()
                result.success(null)
            }
            "seek" -> {
                player.seek(call.argument<Number>("positionMs")?.toLong() ?: 0L, call.argument<Boolean>("precise") ?: true)
                result.success(null)
            }
            "stepFrame" -> {
                player.stepFrame(call.argument<Int>("frames") ?: 1)
                result.success(null)
            }
            "setSpeed" -> {
                player.setSpeed(
                    call.argument<Number>("speed")?.toFloat() ?: 1f,
                    call.argument<Boolean>("preservePitch") ?: true,
                )
                result.success(null)
            }
            "setVolume" -> {
                player.setVolume(call.argument<Number>("volume")?.toFloat() ?: 1f)
                result.success(null)
            }
            "setMuted" -> {
                player.setMuted(call.argument<Boolean>("muted") ?: false)
                result.success(null)
            }
            "setRepeat" -> {
                player.setRepeat(call.argument<String>("mode"))
                result.success(null)
            }
            "selectTrack" -> {
                player.selectTrack(call.argument<String>("trackId") ?: "", call.argument<String>("type"))
                result.success(null)
            }
            "setAutoQuality" -> {
                player.setAutoQuality()
                result.success(null)
            }
            "setMaxHeight" -> {
                player.setMaxHeight(call.argument<Int>("height") ?: 0)
                result.success(null)
            }
            "getEqualizer" -> result.success(player.describeEqualizer())
            "setEqualizerEnabled" -> {
                player.setEqualizerEnabled(call.argument<Boolean>("enabled") ?: false)
                result.success(null)
            }
            "setEqualizerBand" -> {
                player.setEqualizerBand(call.argument<Int>("index") ?: 0, call.argument<Number>("gainDb")?.toDouble() ?: 0.0)
                result.success(null)
            }
            "setEqualizerPreset" -> {
                player.setEqualizerPreset(call.argument<String>("preset") ?: "")
                result.success(null)
            }
            "setBassBoost" -> {
                player.setBassBoost(call.argument<Number>("strength")?.toDouble() ?: 0.0)
                result.success(null)
            }
            "setVirtualizer" -> {
                player.setVirtualizer(call.argument<Number>("strength")?.toDouble() ?: 0.0)
                result.success(null)
            }
            "setLoudness" -> {
                player.setLoudness(call.argument<Number>("gainDb")?.toDouble() ?: 0.0)
                result.success(null)
            }
            "startVisualizer" -> {
                player.startVisualizer(call.argument<Int>("bands") ?: 48)
                result.success(null)
            }
            "stopVisualizer" -> {
                player.stopVisualizer()
                result.success(null)
            }
            "setAudioDelay" -> result.success(null)
            "enterPip" -> result.success(player.enterPip(activity, call.argument<Number>("aspectRatio")?.toDouble() ?: (16.0 / 9.0)))
            "exitPip" -> {
                UMediaPip.exit(activity)
                result.success(null)
            }
            "screenshot" -> result.success(player.screenshot())
            "setNotification" -> result.success(null)
            "dispose" -> {
                player.dispose()
                players.remove(id)
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }
}
